In [1]:
import pandas as pd

### V-dem variables for 2019 year (commented out due to file size)

In [2]:
#vdem_whole = pd.read_csv('vdem_march_2020.csv', low_memory=False)
# vdem = vdem_whole.loc[vdem_whole.year >= 2019]
# vdem.shape

In [3]:
# # there is no data for 2019:
# #'e_wbgi_rqe' (regulatory quality)
# #'e_peedgini' (education variable)
# # 'e_cow_exports', 'e_cow_imports' (export and import)
# vars =['country_name', 'country_text_id', 'year', 'e_regiongeo', 'e_regionpol', 'v2x_rule', 'v2xcl_acjst', 'v2xcl_prpty', 'v2x_polyarchy', 'e_peaveduc']
# for i in vars:
#     print(i, '\n')
#     print(vdem[i].isna().sum())

In [4]:
# vdem = vdem[vars]
# vdem.rename(columns={'e_regiongeo': 'region_geo', 'e_regionpol': 'region_pol', "v2x_polyarchy": 'polyarchy',  'v2x_rule': 'rule', 'v2xcl_acjst': 'justice', 'v2xcl_prpty': 'property', 'e_peaveduc': 'education'}, inplace = True)
# vdem = vdem.sort_index(axis=1, ascending=False)
# vdem.isna().sum()
# vdem.to_csv('vdem_2019.csv')

In [5]:
vdem = pd.read_csv('../../data/main_research/raw_data/vdem_2019.csv')
vdem.drop(columns=['Unnamed: 0'], inplace=True)

In [6]:
vdem.head()

year   rule  region_pol  region_geo  property  polyarchy  justice  \
0  2019  0.605           2          17     0.872      0.710    0.354   
1  2019  0.562          10          18     0.750      0.736    0.764   
2  2019  0.991           5           2     0.892      0.874    0.972   
3  2019  0.993           5           1     0.877      0.873    0.979   
4  2019  0.649           4           6     0.817      0.718    0.891   

   education country_text_id country_name  
0       8.48             MEX       Mexico  
1        NaN             SUR     Suriname  
2      11.85             SWE       Sweden  
3      12.71             CHE  Switzerland  
4       5.61             GHA        Ghana

### Covid variables

#### V-dem data

In [7]:
data = pd.read_csv('../../data/main_research/raw_data/vdem_pandem_data_v3.csv')

In [8]:
data.head()

country_name        time  emlaw  \
0  Afghanistan  2020-06-17      5   
1      Albania  2020-06-08      3   
2      Algeria  2020-06-22      4   
3       Angola  2020-06-15      1   
4    Argentina  2020-06-09      4   

                                              emname     emstart       emend  \
0                                                NaN         NaN         NaN   
1  Constitution (Articles 100, 170, 174, point 1,...  2020-03-24         NaN   
2  Constitution 1989 (reinst. 1996, rev. 2016), a...  2020-03-21         NaN   
3           Constitution Article 58 and Law n. 17/91  2020-03-27  2020-05-25   
4  Ley 27541 - LEY DE SOLIDARIDAD SOCIAL Y REACTI...  2020-03-20         NaN   

   emlimit emlimitdate  emlawapp emlegapp  ...   panback  pandem_dis  \
0      NaN         NaN       NaN      NaN  ...  0.000000        0.05   
1      1.0  2020-06-23       1.0        1  ...  0.173110        0.15   
2      1.0  2020-04-04       0.0        0  ...  0.123340        0.35   
3      1.0  2020-04-11       1.0        1  ...  0.041671        0.10   
4      1.0  2020-03-31       1.0        1  ...  0.054786        0.05   

   pandem_dis_alt  type1  type2 type3 type4  type4_alt  type5  type6  
0            0.00      0      0     0     1          0      0      0  
1            0.15      0      0     0     3          3      0      0  
2            0.35      0      0     0     3          3      2      0  
3            0.05      0      0     0     1          0      0      1  
4            0.05      0      0     0     0          0      0      1  

[5 rows x 41 columns]

In [9]:
data = data[['country_name','pandem','pandem_dis','panback']]

In [10]:
data.head()

country_name    pandem  pandem_dis   panback
0  Afghanistan  0.058824        0.05  0.000000
1      Albania  0.176471        0.15  0.173110
2      Algeria  0.294118        0.35  0.123340
3       Angola  0.117647        0.10  0.041671
4    Argentina  0.058824        0.05  0.054786

#### Adding continents and sub-region columns to dataframe

https://github.com/lukes/ISO-3166-Countries-with-Regional-Codes/blob/master/all/all.csv

In [11]:
continents = pd.read_csv('../../data/main_research/raw_data/country_location.csv')

In [12]:
continents.head()

name alpha-2 alpha-3  country-code     iso_3166-2   region  \
0     Afghanistan      AF     AFG             4  ISO 3166-2:AF     Asia   
1   Åland Islands      AX     ALA           248  ISO 3166-2:AX   Europe   
2         Albania      AL     ALB             8  ISO 3166-2:AL   Europe   
3         Algeria      DZ     DZA            12  ISO 3166-2:DZ   Africa   
4  American Samoa      AS     ASM            16  ISO 3166-2:AS  Oceania   

        sub-region intermediate-region  region-code  sub-region-code  \
0    Southern Asia                 NaN        142.0             34.0   
1  Northern Europe                 NaN        150.0            154.0   
2  Southern Europe                 NaN        150.0             39.0   
3  Northern Africa                 NaN          2.0             15.0   
4        Polynesia                 NaN          9.0             61.0   

   intermediate-region-code  
0                       NaN  
1                       NaN  
2                       NaN  
3                       NaN  
4                       NaN

In [13]:
continents = continents[['name','alpha-3','region','sub-region']]

In [14]:
continents.rename(columns={"name": "country_name", "alpha-3": "country_txt", "region": "continent"}, inplace=True)

In [15]:
df = pd.merge(data, continents, how='left', on=['country_name'])

In [16]:
df.head()

country_name    pandem  pandem_dis   panback country_txt continent  \
0  Afghanistan  0.058824        0.05  0.000000         AFG      Asia   
1      Albania  0.176471        0.15  0.173110         ALB    Europe   
2      Algeria  0.294118        0.35  0.123340         DZA    Africa   
3       Angola  0.117647        0.10  0.041671         AGO    Africa   
4    Argentina  0.058824        0.05  0.054786         ARG  Americas   

                        sub-region  
0                    Southern Asia  
1                  Southern Europe  
2                  Northern Africa  
3               Sub-Saharan Africa  
4  Latin America and the Caribbean

In [17]:
len(data)

146

In [18]:
len(df)

146

#### Checking which countries names are different in both datasets and then updating their names

In [19]:
df1 = df[df.isna().any(axis=1)]
df1['country_name']

13                              Bolivia
19                        Burma/Myanmar
32                       Czech Republic
33     Democratic Republic of the Congo
55                                 Iran
60                          Ivory Coast
68                                 Laos
80                              Moldova
91                          North Korea
104               Republic of the Congo
106                              Russia
117                         South Korea
124                               Syria
125                              Taiwan
127                            Tanzania
129                          The Gambia
137                      United Kingdom
141                           Venezuela
142                             Vietnam
Name: country_name, dtype: object

In [20]:
continents.loc[234:244]

country_name country_txt continent  \
234  United Kingdom of Great Britain and Northern I...         GBR    Europe   
235                           United States of America         USA  Americas   
236               United States Minor Outlying Islands         UMI   Oceania   
237                                            Uruguay         URY  Americas   
238                                         Uzbekistan         UZB      Asia   
239                                            Vanuatu         VUT   Oceania   
240                 Venezuela (Bolivarian Republic of)         VEN  Americas   
241                                           Viet Nam         VNM      Asia   
242                           Virgin Islands (British)         VGB  Americas   
243                              Virgin Islands (U.S.)         VIR  Americas   
244                                  Wallis and Futuna         WLF   Oceania   

                          sub-region  
234                  Northern Europe  
235                 Northern America  
236                       Micronesia  
237  Latin America and the Caribbean  
238                     Central Asia  
239                        Melanesia  
240  Latin America and the Caribbean  
241               South-eastern Asia  
242  Latin America and the Caribbean  
243  Latin America and the Caribbean  
244                        Polynesia

In [21]:
update = {
    'Bolivia (Plurinational State of)': 'Bolivia',
    'Czechia': 'Czech Republic',
    "Côte d'Ivoire": 'Ivory Coast',
    'Congo, Democratic Republic of the': 'Democratic Republic of the Congo',
    'Congo': 'Republic of the Congo',
    'Gambia': 'The Gambia',
    'Iran (Islamic Republic of)': 'Iran',
    "Korea (Democratic People's Republic of)": 'North Korea',
    'Korea, Republic of': 'South Korea',
    "Lao People's Democratic Republic": 'Laos',
    'Moldova, Republic of': 'Moldova',
    'Myanmar': 'Burma/Myanmar',
    'Russian Federation': 'Russia',
    'Syrian Arab Republic': 'Syria',
    'Taiwan, Province of China': 'Taiwan',
    'Tanzania, United Republic of': 'Tanzania',
    'Venezuela (Bolivarian Republic of)': 'Venezuela',
    'Viet Nam': 'Vietnam',
    'United Kingdom of Great Britain and Northern Ireland': 'United Kingdom'
}

In [22]:
continents.replace({"country_name": update}, inplace=True)

In [23]:
df = pd.merge(data, continents, how='left', on=['country_name'])

df.head()

country_name    pandem  pandem_dis   panback country_txt continent  \
0  Afghanistan  0.058824        0.05  0.000000         AFG      Asia   
1      Albania  0.176471        0.15  0.173110         ALB    Europe   
2      Algeria  0.294118        0.35  0.123340         DZA    Africa   
3       Angola  0.117647        0.10  0.041671         AGO    Africa   
4    Argentina  0.058824        0.05  0.054786         ARG  Americas   

                        sub-region  
0                    Southern Asia  
1                  Southern Europe  
2                  Northern Africa  
3               Sub-Saharan Africa  
4  Latin America and the Caribbean

In [24]:
df['country_txt'].isna().sum()

0

#### Adding income group according to World Bank classification

https://datahelpdesk.worldbank.org/knowledgebase/articles/906519-world-bank-country-and-lending-groups

In [25]:
classification = pd.read_excel('../../data/main_research/raw_data/country_classification_world_bank.xls', skiprows=4)
classification.head()

x  x.1         Economy Code    X                      Region  \
0  x    x               x    x    x                           x   
1  1  NaN     Afghanistan  AFG  NaN                  South Asia   
2  2  NaN         Albania  ALB  NaN       Europe & Central Asia   
3  3  NaN         Algeria  DZA  NaN  Middle East & North Africa   
4  4  NaN  American Samoa  ASM  NaN         East Asia & Pacific   

          Income group Lending category Other  
0                    x                x     x  
1           Low income              IDA  HIPC  
2  Upper middle income             IBRD   NaN  
3  Lower middle income             IBRD   NaN  
4  Upper middle income               ..   NaN

In [26]:
classification = classification[['Code', 'Income group']].iloc[1:219]
classification.rename(columns={"Code": "country_txt", "Income group": "income_group"}, inplace=True)
classification.tail()

country_txt         income_group
214         VIR          High income
215         PSE  Lower middle income
216         YEM           Low income
217         ZMB  Lower middle income
218         ZWE  Lower middle income

In [27]:
# Final merging
df = pd.merge(df, classification, how='left', on=['country_txt'])

In [28]:
df.to_csv('pandem.csv', index=False)

In [29]:
del df
del data

In [30]:
pandem = pd.read_csv('pandem.csv')

In [31]:
pandem.head()

country_name    pandem  pandem_dis   panback country_txt continent  \
0  Afghanistan  0.058824        0.05  0.000000         AFG      Asia   
1      Albania  0.176471        0.15  0.173110         ALB    Europe   
2      Algeria  0.294118        0.35  0.123340         DZA    Africa   
3       Angola  0.117647        0.10  0.041671         AGO    Africa   
4    Argentina  0.058824        0.05  0.054786         ARG  Americas   

                        sub-region         income_group  
0                    Southern Asia           Low income  
1                  Southern Europe  Upper middle income  
2                  Northern Africa  Lower middle income  
3               Sub-Saharan Africa  Lower middle income  
4  Latin America and the Caribbean  Upper middle income

In [32]:
pandem.rename(columns={'country_txt': 'country_text_id'}, inplace = True)

### Checking country_name compliance in vdem and pandem datasets

In [33]:
count = vdem[['country_name', 'country_text_id']]
countries_df = count.groupby('country_name')['country_text_id'].max().to_frame().reset_index().sort_values(by='country_name')

count = pandem[['country_name', 'country_text_id']]
countries_df_pandem = count.groupby('country_text_id')['country_name'].max().to_frame().reset_index().sort_values(by='country_text_id')


result = pd.merge(countries_df, countries_df_pandem, how='left', on=['country_text_id'])

result['differences'] = result['country_name_x'] == result['country_name_y']
good = list(result[result['differences'] == True].country_text_id)
missings = list(result[result['differences'] == False].country_text_id)
print(len(good))
print(pandem.shape[0])

146
146


In [34]:
vdem.rename(columns={'country_name': 'country'}, inplace = True)

#### Merging two datasets on 'country_text_id'

In [35]:
data = pd.merge(pandem, vdem, how='left', on=['country_text_id'])

In [36]:
data.head()

country_name    pandem  pandem_dis   panback country_text_id continent  \
0  Afghanistan  0.058824        0.05  0.000000             AFG      Asia   
1      Albania  0.176471        0.15  0.173110             ALB    Europe   
2      Algeria  0.294118        0.35  0.123340             DZA    Africa   
3       Angola  0.117647        0.10  0.041671             AGO    Africa   
4    Argentina  0.058824        0.05  0.054786             ARG  Americas   

                        sub-region         income_group  year   rule  \
0                    Southern Asia           Low income  2019  0.341   
1                  Southern Europe  Upper middle income  2019  0.596   
2                  Northern Africa  Lower middle income  2019  0.217   
3               Sub-Saharan Africa  Lower middle income  2019  0.406   
4  Latin America and the Caribbean  Upper middle income  2019  0.757   

   region_pol  region_geo  property  polyarchy  justice  education  \
0           8          14     0.423      0.347    0.198      3.813   
1           1           3     0.737      0.481    0.936        NaN   
2           3           5     0.611      0.300    0.500      7.300   
3           4           7     0.603      0.386    0.487      2.980   
4           2          18     0.824      0.812    0.707      8.770   

       country  
0  Afghanistan  
1      Albania  
2      Algeria  
3       Angola  
4    Argentina

In [37]:
data.drop(columns=['country'], inplace=True)

In [38]:
data.isna().sum()

country_name        0
pandem              0
pandem_dis          0
panback             0
country_text_id     0
continent           0
sub-region          0
income_group        0
year                0
rule                0
region_pol          0
region_geo          0
property            0
polyarchy           0
justice             0
education          20
dtype: int64

### World Bank variables

#### Shortening dataset to variables for 2010-2019

In [39]:
names = ['population_density_world_bank', 'export_gdp_world_bank', 'gdp_pc_world_bank', 'gini_index_world_bank', 
         'import_gdp_world_bank', 'inflation_world_bank', 'military_world_bank', 'mineral_world_bank', 'oil_world_bank']
for i in range(len(names)):
    df = pd.read_csv(f'../../data/main_research/raw_data/{names[i]}.csv', sep=',', skiprows=4)
    df.drop(columns = ['Indicator Code', 'Indicator Name', 'Unnamed: 64'], inplace=True)
    reshaped_df = df.melt(id_vars=['Country Name', 'Country Code'])
    reshaped_df.rename(columns={ reshaped_df.columns[0]: "country_name", reshaped_df.columns[1]: "country_text_id", reshaped_df.columns[3]: f'{names[i]}', reshaped_df.columns[2]: "year" }, inplace = True)
    reshaped_df = reshaped_df.loc[(reshaped_df.year).astype(int) >= 2010]
    if(i == 0):
        WB = reshaped_df.copy()
    else:
        WB[f'{names[i]}'] = reshaped_df[f'{names[i]}']

In [40]:
WB.head()

country_name country_text_id  year  population_density_world_bank  \
13200        Aruba             ABW  2010                     564.827778   
13201  Afghanistan             AFG  2010                      44.704082   
13202       Angola             AGO  2010                      18.734456   
13203      Albania             ALB  2010                     106.314635   
13204      Andorra             AND  2010                     179.678723   

       export_gdp_world_bank  gdp_pc_world_bank  gini_index_world_bank  \
13200              61.042300       23512.602596                    NaN   
13201              16.182633         543.303042                    NaN   
13202              61.543113        3587.883798                    NaN   
13203              27.979423        4094.350334                    NaN   
13204                    NaN       40852.666777                    NaN   

       import_gdp_world_bank  inflation_world_bank  military_world_bank  \
13200              77.564852              2.078141                  NaN   
13201              36.396967              2.178538             1.945837   
13202              42.580517            -60.496401             4.177585   
13203              48.563932              3.622335             1.558592   
13204                    NaN                   NaN                  NaN   

       mineral_world_bank  oil_world_bank  
13200            0.000000        0.000000  
13201            0.000000        0.004974  
13202            0.000000       38.534368  
13203            0.462026        1.499436  
13204            0.000000             NaN

In [41]:
WB.isna().sum()

country_name                        0
country_text_id                     0
year                                0
population_density_world_bank     301
export_gdp_world_bank             392
gdp_pc_world_bank                 169
gini_index_world_bank            1996
import_gdp_world_bank             392
inflation_world_bank              405
military_world_bank               897
mineral_world_bank                659
oil_world_bank                    744
dtype: int64

#### Filling 'missing values' with the latest value for each country

In [42]:
for i in names:
    WB[i] = WB.groupby('country_name')[i].fillna(method='ffill')

In [43]:
WB = WB.loc[(WB.year).astype(int) >= 2019]
WB.head()

country_name country_text_id  year  population_density_world_bank  \
15576        Aruba             ABW  2019                     588.027778   
15577  Afghanistan             AFG  2019                      56.937760   
15578       Angola             AGO  2019                      24.713052   
15579      Albania             ALB  2019                     104.612263   
15580      Andorra             AND  2019                     163.842553   

       export_gdp_world_bank  gdp_pc_world_bank  gini_index_world_bank  \
15576              73.332115       29007.693003                    NaN   
15577               7.753798         502.115487                    NaN   
15578              40.836290        2973.591160                   51.3   
15579              31.511223        5352.857411                   33.2   
15580                    NaN       40886.391165                    NaN   

       import_gdp_world_bank  inflation_world_bank  military_world_bank  \
15576              71.632243              4.257462                  NaN   
15577              37.867379              2.302373             0.984561   
15578              25.541722             17.145320             1.777138   
15579              45.217077              1.411091             1.178901   
15580                    NaN                   NaN                  NaN   

       mineral_world_bank  oil_world_bank  
15576            0.000000        0.000000  
15577            0.007421        0.001255  
15578            0.000000       15.753913  
15579            0.146426        1.365752  
15580            0.000000             NaN

In [44]:
WB.isna().sum()

country_name                       0
country_text_id                    0
year                               0
population_density_world_bank      3
export_gdp_world_bank             25
gdp_pc_world_bank                 11
gini_index_world_bank            115
import_gdp_world_bank             25
inflation_world_bank              30
military_world_bank               59
mineral_world_bank                13
oil_world_bank                    24
dtype: int64

#### Filling 'missing values' with 0 for 'Oil' and 'Mineral' variables (The assumption that in the absence of complete data for 2010-2019 - the country does not produce oil / mineral)

In [45]:
WB['oil_world_bank'] = WB['oil_world_bank'].fillna(0)
WB['mineral_world_bank'] = WB['mineral_world_bank'].fillna(0)

In [46]:
WB.isna().sum()

country_name                       0
country_text_id                    0
year                               0
population_density_world_bank      3
export_gdp_world_bank             25
gdp_pc_world_bank                 11
gini_index_world_bank            115
import_gdp_world_bank             25
inflation_world_bank              30
military_world_bank               59
mineral_world_bank                 0
oil_world_bank                     0
dtype: int64

### Checking country name compliance in both datasets

In [47]:
count = WB[['country_name', 'country_text_id']]
countries_df = count.groupby('country_name')['country_text_id'].max().to_frame().reset_index().sort_values(by='country_name')

count = pandem[['country_name', 'country_text_id']]
countries_df_pandem = count.groupby('country_text_id')['country_name'].max().to_frame().reset_index().sort_values(by='country_text_id')


result = pd.merge(countries_df_pandem, countries_df, how='left', on=['country_text_id'])

result['differences'] = result['country_name_x'] == result['country_name_y']
good = list(result[result['differences'] == True].country_text_id)
missings = list(result[result['differences'] == False].country_text_id)
print(len(good))
print(pandem.shape[0])

128
146


In [48]:
result.tail()

country_text_id country_name_x country_name_y  differences
141             VNM        Vietnam        Vietnam         True
142             YEM          Yemen    Yemen, Rep.        False
143             ZAF   South Africa   South Africa         True
144             ZMB         Zambia         Zambia         True
145             ZWE       Zimbabwe       Zimbabwe         True

In [49]:
result.isna().sum()
## Only one country without 'country_text_id' in World Bank database

country_text_id    0
country_name_x     0
country_name_y     1
differences        0
dtype: int64

In [50]:
result.loc[result['country_name_y'].isna()]

country_text_id country_name_x country_name_y  differences
133             TWN         Taiwan            NaN        False

In [51]:
WB.loc[WB.country_name.str.contains('Taiwan')]

Empty DataFrame
Columns: [country_name, country_text_id, year, population_density_world_bank, export_gdp_world_bank, gdp_pc_world_bank, gini_index_world_bank, import_gdp_world_bank, inflation_world_bank, military_world_bank, mineral_world_bank, oil_world_bank]
Index: []

https://datahelpdesk.worldbank.org/knowledgebase/articles/114933-where-are-your-data-on-taiwan

#### Taiwan, China, is not listed as a separate country for World Development Indicators

In [52]:
WB.drop(columns=['country_name'], inplace=True)
WB.drop(columns=['year'], inplace=True)
data = pd.merge(data, WB, how='left', on=['country_text_id'])

In [53]:
data.head()

country_name    pandem  pandem_dis   panback country_text_id continent  \
0  Afghanistan  0.058824        0.05  0.000000             AFG      Asia   
1      Albania  0.176471        0.15  0.173110             ALB    Europe   
2      Algeria  0.294118        0.35  0.123340             DZA    Africa   
3       Angola  0.117647        0.10  0.041671             AGO    Africa   
4    Argentina  0.058824        0.05  0.054786             ARG  Americas   

                        sub-region         income_group  year   rule  ...  \
0                    Southern Asia           Low income  2019  0.341  ...   
1                  Southern Europe  Upper middle income  2019  0.596  ...   
2                  Northern Africa  Lower middle income  2019  0.217  ...   
3               Sub-Saharan Africa  Lower middle income  2019  0.406  ...   
4  Latin America and the Caribbean  Upper middle income  2019  0.757  ...   

   education  population_density_world_bank  export_gdp_world_bank  \
0      3.813                      56.937760               7.753798   
1        NaN                     104.612263              31.511223   
2      7.300                      17.730075              22.570151   
3      2.980                      24.713052              40.836290   
4      8.770                      16.258510              17.290218   

   gdp_pc_world_bank  gini_index_world_bank  import_gdp_world_bank  \
0         502.115487                    NaN              37.867379   
1        5352.857411                   33.2              45.217077   
2        3948.343279                   27.6              29.474623   
3        2973.591160                   51.3              25.541722   
4       10006.148974                   41.4              15.103448   

   inflation_world_bank  military_world_bank  mineral_world_bank  \
0              2.302373             0.984561            0.007421   
1              1.411091             1.178901            0.146426   
2              1.951768             5.271414            0.046717   
3             17.145320             1.777138            0.000000   
4             53.548304             0.854561            0.308310   

   oil_world_bank  
0        0.001255  
1        1.365752  
2       12.313697  
3       15.753913  
4        0.735824  

[5 rows x 25 columns]

### Fractionalization dataset

In [54]:
frac = pd.read_excel('../../data/main_research/raw_data/fractionalization.xlsx')

In [55]:
frac.head()

country_name  ethnic_frac  ling_frac  relig_frac
0     Afghanistan       0.7693     0.6141      0.2717
1         Albania       0.2204     0.0400      0.4719
2         Algeria       0.3394     0.4427      0.0100
3  American Samoa       0.0000     0.1733      0.6395
4         Andorra       0.7139     0.6848      0.2326

In [56]:
countries_frac = list(frac['country_name'])
countries_pandem = list(pandem['country_name'])

In [57]:
sum([item in countries_pandem for item in countries_frac])

135

In [58]:
sum([item in countries_frac for item in countries_pandem])

135

#### 'country_name' complies in 135 cases!

#### Trial merging to check which 'country_name' differs between datasets

In [59]:
df = pd.merge(data[['country_name']], frac, how='left', on=['country_name'])

#### Checking which countries names are different in both datasets and then updating their names

In [60]:
df1 = df[df.isna().any(axis=1)]
df1['country_name']

19                        Burma/Myanmar
33     Democratic Republic of the Congo
91                          North Korea
92                      North Macedonia
104               Republic of the Congo
106                              Russia
113                            Slovakia
117                         South Korea
118                         South Sudan
129                          The Gambia
138            United States of America
Name: country_name, dtype: object

In [61]:
frac.loc[180:185]

country_name  ethnic_frac  ling_frac  relig_frac
180              Sri Lanka       0.4150     0.4645      0.4853
181  Saint Kitts and Nevis       0.1842     0.0000      0.6614
182                  Sudan       0.7147     0.7190      0.4307
183               Suriname       0.7332     0.3310      0.7910
184              Swaziland       0.0580     0.1722      0.4444
185                 Sweden       0.1000     0.1968      0.2342

In [62]:
frac.loc[frac.country_name.str.contains('Sudan')]

country_name  ethnic_frac  ling_frac  relig_frac
182        Sudan       0.7147      0.719      0.4307

#### Dictionary for updating country names (South Sudan does not exist in 2003)

In [63]:
update = {
    'Burma': 'Burma/Myanmar',
    'Democratic Republic of Congo': 'Democratic Republic of the Congo',
    'Congo': 'Republic of the Congo',
    'Gambia, The': 'The Gambia',
    "Korea, North": 'North Korea',
    'Korea, South': 'South Korea',
    'Russian Federation': 'Russia',
    'Macedonia': 'North Macedonia',
    'Slovak Republic': 'Slovakia',
    #'South Sudan'
    'United States': 'United States of America'
}

In [64]:
frac.replace({"country_name": update}, inplace=True)

### Final merging

In [65]:
data = pd.merge(data, frac, how='left', on=['country_name'])

### Full dataset with some missings

In [66]:
# data.drop(columns=['year'], inplace=True)
data['trade_gdp_world_bank'] = data['export_gdp_world_bank'] + data['import_gdp_world_bank']
data.drop(columns=['export_gdp_world_bank', 'import_gdp_world_bank'], inplace=True)
data.head()

country_name    pandem  pandem_dis   panback country_text_id continent  \
0  Afghanistan  0.058824        0.05  0.000000             AFG      Asia   
1      Albania  0.176471        0.15  0.173110             ALB    Europe   
2      Algeria  0.294118        0.35  0.123340             DZA    Africa   
3       Angola  0.117647        0.10  0.041671             AGO    Africa   
4    Argentina  0.058824        0.05  0.054786             ARG  Americas   

                        sub-region         income_group  year   rule  ...  \
0                    Southern Asia           Low income  2019  0.341  ...   
1                  Southern Europe  Upper middle income  2019  0.596  ...   
2                  Northern Africa  Lower middle income  2019  0.217  ...   
3               Sub-Saharan Africa  Lower middle income  2019  0.406  ...   
4  Latin America and the Caribbean  Upper middle income  2019  0.757  ...   

   gdp_pc_world_bank  gini_index_world_bank  inflation_world_bank  \
0         502.115487                    NaN              2.302373   
1        5352.857411                   33.2              1.411091   
2        3948.343279                   27.6              1.951768   
3        2973.591160                   51.3             17.145320   
4       10006.148974                   41.4             53.548304   

   military_world_bank  mineral_world_bank  oil_world_bank  ethnic_frac  \
0             0.984561            0.007421        0.001255       0.7693   
1             1.178901            0.146426        1.365752       0.2204   
2             5.271414            0.046717       12.313697       0.3394   
3             1.777138            0.000000       15.753913       0.7867   
4             0.854561            0.308310        0.735824       0.2550   

   ling_frac  relig_frac  trade_gdp_world_bank  
0     0.6141      0.2717             45.621177  
1     0.0400      0.4719             76.728300  
2     0.4427      0.0100             52.044775  
3     0.7870      0.6276             66.378012  
4     0.0620      0.2236             32.393666  

[5 rows x 27 columns]

### Missing values

In [67]:
data.isna().sum()

country_name                      0
pandem                            0
pandem_dis                        0
panback                           0
country_text_id                   0
continent                         0
sub-region                        0
income_group                      0
year                              0
rule                              0
region_pol                        0
region_geo                        0
property                          0
polyarchy                         0
justice                           0
education                        20
population_density_world_bank     3
gdp_pc_world_bank                 4
gini_index_world_bank            25
inflation_world_bank              8
military_world_bank               5
mineral_world_bank                1
oil_world_bank                    1
ethnic_frac                       1
ling_frac                         1
relig_frac                        1
trade_gdp_world_bank              6
dtype: int64

### Data imputation

#### Fractionalization for South Sudan
In 2003 South Sudan was a part of Sudan, so we impute for South Sudan values for Sudan

In [68]:
data.loc[data['country_name'] == 'South Sudan','ethnic_frac'] = data.loc[data['country_name'] == 'Sudan']['ethnic_frac'].values[0]
data.loc[data['country_name'] == 'South Sudan','ling_frac'] = data.loc[data['country_name'] == 'Sudan']['ling_frac'].values[0]
data.loc[data['country_name'] == 'South Sudan','relig_frac'] = data.loc[data['country_name'] == 'Sudan']['relig_frac'].values[0]

In [69]:
data.loc[data['country_name'] == 'South Sudan']

country_name    pandem  pandem_dis  panback country_text_id continent  \
118  South Sudan  0.176471        0.15      0.0             SSD    Africa   

             sub-region income_group  year   rule  ...  gdp_pc_world_bank  \
118  Sub-Saharan Africa   Low income  2019  0.051  ...        1119.651333   

     gini_index_world_bank  inflation_world_bank  military_world_bank  \
118                    NaN            187.851626             1.329208   

     mineral_world_bank  oil_world_bank  ethnic_frac  ling_frac  relig_frac  \
118                 0.0       31.325699       0.7147      0.719      0.4307   

     trade_gdp_world_bank  
118             65.551349  

[1 rows x 27 columns]

In [70]:
data.loc[data['relig_frac'].isna()]

Empty DataFrame
Columns: [country_name, pandem, pandem_dis, panback, country_text_id, continent, sub-region, income_group, year, rule, region_pol, region_geo, property, polyarchy, justice, education, population_density_world_bank, gdp_pc_world_bank, gini_index_world_bank, inflation_world_bank, military_world_bank, mineral_world_bank, oil_world_bank, ethnic_frac, ling_frac, relig_frac, trade_gdp_world_bank]
Index: []

[0 rows x 27 columns]

### 'Oil' and 'Mineral' for Taiwan
"Taiwan produces a small amount of natural gas." We impute 0 for Oil and Mineral for simplicity.

In [71]:
data.loc[data['country_name'] == 'Taiwan']

country_name    pandem  pandem_dis   panback country_text_id continent  \
125       Taiwan  0.058824        0.05  0.049692             TWN      Asia   

       sub-region income_group  year   rule  ...  gdp_pc_world_bank  \
125  Eastern Asia  High income  2019  0.913  ...                NaN   

     gini_index_world_bank  inflation_world_bank  military_world_bank  \
125                    NaN                   NaN                  NaN   

     mineral_world_bank  oil_world_bank  ethnic_frac  ling_frac  relig_frac  \
125                 NaN             NaN       0.2744     0.5028      0.6845   

     trade_gdp_world_bank  
125                   NaN  

[1 rows x 27 columns]

In [72]:
data.loc[data['country_name'] == 'Taiwan','oil_world_bank'] = 0
data.loc[data['country_name'] == 'Taiwan','mineral_world_bank'] = 0

In [73]:
data.loc[data['country_name'] == 'Taiwan']

country_name    pandem  pandem_dis   panback country_text_id continent  \
125       Taiwan  0.058824        0.05  0.049692             TWN      Asia   

       sub-region income_group  year   rule  ...  gdp_pc_world_bank  \
125  Eastern Asia  High income  2019  0.913  ...                NaN   

     gini_index_world_bank  inflation_world_bank  military_world_bank  \
125                    NaN                   NaN                  NaN   

     mineral_world_bank  oil_world_bank  ethnic_frac  ling_frac  relig_frac  \
125                 0.0             0.0       0.2744     0.5028      0.6845   

     trade_gdp_world_bank  
125                   NaN  

[1 rows x 27 columns]

### Imputation of 'Density'.

Wikipedia as a source.

In [74]:
data.loc[data['population_density_world_bank'].isna()]

country_name    pandem  pandem_dis   panback country_text_id continent  \
118  South Sudan  0.176471        0.15  0.000000             SSD    Africa   
121        Sudan  0.176471        0.15  0.056069             SDN    Africa   
125       Taiwan  0.058824        0.05  0.049692             TWN      Asia   

             sub-region income_group  year   rule  ...  gdp_pc_world_bank  \
118  Sub-Saharan Africa   Low income  2019  0.051  ...        1119.651333   
121     Northern Africa   Low income  2019  0.269  ...         441.505603   
125        Eastern Asia  High income  2019  0.913  ...                NaN   

     gini_index_world_bank  inflation_world_bank  military_world_bank  \
118                    NaN            187.851626             1.329208   
121                   34.2             50.994052             2.278302   
125                    NaN                   NaN                  NaN   

     mineral_world_bank  oil_world_bank  ethnic_frac  ling_frac  relig_frac  \
118            0.000000       31.325699       0.7147     0.7190      0.4307   
121            2.537048        1.013017       0.7147     0.7190      0.4307   
125            0.000000        0.000000       0.2744     0.5028      0.6845   

     trade_gdp_world_bank  
118             65.551349  
121             26.721863  
125                   NaN  

[3 rows x 27 columns]

In [75]:
data.loc[data['country_name'] == 'Taiwan','population_density_world_bank'] = 650
data.loc[data['country_name'] == 'Sudan','population_density_world_bank'] = 21.3
data.loc[data['country_name'] == 'South Sudan','population_density_world_bank'] = 13.33

In [76]:
data.loc[data['population_density_world_bank'].isna()]

Empty DataFrame
Columns: [country_name, pandem, pandem_dis, panback, country_text_id, continent, sub-region, income_group, year, rule, region_pol, region_geo, property, polyarchy, justice, education, population_density_world_bank, gdp_pc_world_bank, gini_index_world_bank, inflation_world_bank, military_world_bank, mineral_world_bank, oil_world_bank, ethnic_frac, ling_frac, relig_frac, trade_gdp_world_bank]
Index: []

[0 rows x 27 columns]

In [77]:
data.isna().sum()

country_name                      0
pandem                            0
pandem_dis                        0
panback                           0
country_text_id                   0
continent                         0
sub-region                        0
income_group                      0
year                              0
rule                              0
region_pol                        0
region_geo                        0
property                          0
polyarchy                         0
justice                           0
education                        20
population_density_world_bank     0
gdp_pc_world_bank                 4
gini_index_world_bank            25
inflation_world_bank              8
military_world_bank               5
mineral_world_bank                0
oil_world_bank                    0
ethnic_frac                       0
ling_frac                         0
relig_frac                        0
trade_gdp_world_bank              6
dtype: int64

### Imputation of 'GDPpc'

Wikipedia as a source

In [78]:
data.loc[data['gdp_pc_world_bank'].isna()]

country_name    pandem  pandem_dis   panback country_text_id continent  \
91   North Korea  0.176471        0.15  0.000000             PRK      Asia   
115      Somalia  0.352941        0.30  0.129312             SOM    Africa   
124        Syria  0.235294        0.25  0.038730             SYR      Asia   
125       Taiwan  0.058824        0.05  0.049692             TWN      Asia   

             sub-region income_group  year   rule  ...  gdp_pc_world_bank  \
91         Eastern Asia   Low income  2019  0.044  ...                NaN   
115  Sub-Saharan Africa   Low income  2019  0.109  ...                NaN   
124        Western Asia   Low income  2019  0.091  ...                NaN   
125        Eastern Asia  High income  2019  0.913  ...                NaN   

     gini_index_world_bank  inflation_world_bank  military_world_bank  \
91                     NaN                   NaN                  NaN   
115                    NaN                   NaN                  NaN   
124                    NaN             36.702295             4.056142   
125                    NaN                   NaN                  NaN   

     mineral_world_bank  oil_world_bank  ethnic_frac  ling_frac  relig_frac  \
91                  0.0             0.0       0.0000     0.0000      0.6604   
115                 0.0             0.0       0.8117     0.0330      0.0000   
124                 0.0             0.0       0.5399     0.1817      0.4310   
125                 0.0             0.0       0.2744     0.5028      0.6845   

     trade_gdp_world_bank  
91                    NaN  
115                   NaN  
124                   NaN  
125                   NaN  

[4 rows x 27 columns]

In [79]:
data.loc[data['country_name'] == 'North Korea','gdp_pc_world_bank'] = 1300
data.loc[data['country_name'] == 'Somalia','gdp_pc_world_bank'] = 348
data.loc[data['country_name'] == 'Syria','gdp_pc_world_bank'] = 831
data.loc[data['country_name'] == 'Taiwan','gdp_pc_world_bank'] = 24828

In [80]:
data.loc[data['gdp_pc_world_bank'].isna()]

Empty DataFrame
Columns: [country_name, pandem, pandem_dis, panback, country_text_id, continent, sub-region, income_group, year, rule, region_pol, region_geo, property, polyarchy, justice, education, population_density_world_bank, gdp_pc_world_bank, gini_index_world_bank, inflation_world_bank, military_world_bank, mineral_world_bank, oil_world_bank, ethnic_frac, ling_frac, relig_frac, trade_gdp_world_bank]
Index: []

[0 rows x 27 columns]

In [81]:
data.isna().sum()

country_name                      0
pandem                            0
pandem_dis                        0
panback                           0
country_text_id                   0
continent                         0
sub-region                        0
income_group                      0
year                              0
rule                              0
region_pol                        0
region_geo                        0
property                          0
polyarchy                         0
justice                           0
education                        20
population_density_world_bank     0
gdp_pc_world_bank                 0
gini_index_world_bank            25
inflation_world_bank              8
military_world_bank               5
mineral_world_bank                0
oil_world_bank                    0
ethnic_frac                       0
ling_frac                         0
relig_frac                        0
trade_gdp_world_bank              6
dtype: int64

### Imputation of 'Gini'
* from Wikipedia
* mean according to income group for unavailable data

In [82]:
data.loc[data['gini_index_world_bank'].isna()]['country_name'].to_list()

['Afghanistan',
 'Azerbaijan',
 'Cambodia',
 'Central African Republic',
 'Cuba',
 'Eritrea',
 'Jamaica',
 'Kuwait',
 'Libya',
 'Mali',
 'New Zealand',
 'Nigeria',
 'North Korea',
 'Oman',
 'Papua New Guinea',
 'Qatar',
 'Saudi Arabia',
 'Singapore',
 'Somalia',
 'South Sudan',
 'Syria',
 'Taiwan',
 'Turkmenistan',
 'Uzbekistan',
 'Venezuela']

In [83]:
### From Wikipedia:
data.loc[data['country_name'] == 'Afghanistan','gini_index_world_bank'] = 27.8
data.loc[data['country_name'] == 'Azerbaijan','gini_index_world_bank'] = 26.6
data.loc[data['country_name'] == 'Cambodia','gini_index_world_bank'] = 36
data.loc[data['country_name'] == 'Central African Republic','gini_index_world_bank'] = 56.3
data.loc[data['country_name'] == 'Cuba','gini_index_world_bank'] = 38
data.loc[data['country_name'] == 'Jamaica','gini_index_world_bank'] = 35
data.loc[data['country_name'] == 'Mali','gini_index_world_bank'] = 33
data.loc[data['country_name'] == 'New Zealand','gini_index_world_bank'] = 33.9
data.loc[data['country_name'] == 'Nigeria','gini_index_world_bank'] = 35.1
data.loc[data['country_name'] == 'Papua New Guinea','gini_index_world_bank'] = 41.9
data.loc[data['country_name'] == 'Qatar','gini_index_world_bank'] = 41.1
data.loc[data['country_name'] == 'Saudi Arabia','gini_index_world_bank'] = 45.9
data.loc[data['country_name'] == 'Singapore','gini_index_world_bank'] = 45.9
data.loc[data['country_name'] == 'South Sudan','gini_index_world_bank'] = 45.5
data.loc[data['country_name'] == 'Syria','gini_index_world_bank'] = 55.8
data.loc[data['country_name'] == 'Taiwan','gini_index_world_bank'] = 34.1
data.loc[data['country_name'] == 'Turkmenistan','gini_index_world_bank'] = 40.8
data.loc[data['country_name'] == 'Uzbekistan','gini_index_world_bank'] = 36.7
data.loc[data['country_name'] == 'Venezuela','gini_index_world_bank'] = 44.8

#### Countries with missings

In [84]:
to_fill = data.loc[data['gini_index_world_bank'].isna()]['country_name'].to_list()
to_fill

['Eritrea', 'Kuwait', 'Libya', 'North Korea', 'Oman', 'Somalia']

Filling NaN's for 'Gini' with means for income group to which country belongs

In [85]:
data.groupby('income_group')['gini_index_world_bank'].mean()

income_group
High income            33.800000
Low income             40.180000
Lower middle income    38.842105
Upper middle income    40.386842
Name: gini_index_world_bank, dtype: float64

In [86]:
high = data.groupby('income_group')['gini_index_world_bank'].mean()['High income']
low = data.groupby('income_group')['gini_index_world_bank'].mean()['Low income']
lmid = data.groupby('income_group')['gini_index_world_bank'].mean()['Lower middle income']
umid = data.groupby('income_group')['gini_index_world_bank'].mean()['Upper middle income']

In [87]:
for i in to_fill:
    if(data.loc[data['country_name'] == i, 'income_group'].values[0] == 'High income'):
        data.loc[data['country_name'] == i,'gini_index_world_bank'] = high
    if(data.loc[data['country_name'] == i, 'income_group'].values[0] == 'Low income'):
        data.loc[data['country_name'] == i,'gini_index_world_bank'] = low  
    if(data.loc[data['country_name'] == i, 'income_group'].values[0] == 'Lower middle income'):
        data.loc[data['country_name'] == i,'gini_index_world_bank'] = lmid
    else:
        data.loc[data['country_name'] == i,'gini_index_world_bank'] = umid

In [88]:
data.isna().sum()

country_name                      0
pandem                            0
pandem_dis                        0
panback                           0
country_text_id                   0
continent                         0
sub-region                        0
income_group                      0
year                              0
rule                              0
region_pol                        0
region_geo                        0
property                          0
polyarchy                         0
justice                           0
education                        20
population_density_world_bank     0
gdp_pc_world_bank                 0
gini_index_world_bank             0
inflation_world_bank              8
military_world_bank               5
mineral_world_bank                0
oil_world_bank                    0
ethnic_frac                       0
ling_frac                         0
relig_frac                        0
trade_gdp_world_bank              6
dtype: int64

### Imputation of other variables
* mean according to income group

##### Function for imputing missings for concrete variable with means for income group to which country belongs

In [89]:
def mean_filler(var):
    
    to_fill = data.loc[data[var].isna()]['country_name'].to_list()

    high = data.groupby('income_group')[var].mean()['High income']
    low = data.groupby('income_group')[var].mean()['Low income']
    lmid = data.groupby('income_group')[var].mean()['Lower middle income']
    umid = data.groupby('income_group')[var].mean()['Upper middle income']

    for i in to_fill:
        if(data.loc[data['country_name'] == i, 'income_group'].values[0] == 'High income'):
            data.loc[data['country_name'] == i,var] = high
        if(data.loc[data['country_name'] == i, 'income_group'].values[0] == 'Low income'):
            data.loc[data['country_name'] == i,var] = low  
        if(data.loc[data['country_name'] == i, 'income_group'].values[0] == 'Lower middle income'):
            data.loc[data['country_name'] == i,var] = lmid
        else:
            data.loc[data['country_name'] == i,var] = umid

In [90]:
data.loc[data['inflation_world_bank'].isna()]['country_name'].to_list()

['Central African Republic',
 'Cuba',
 'Eritrea',
 'North Korea',
 'Somalia',
 'Taiwan',
 'Turkmenistan',
 'Uzbekistan']

In [91]:
mean_filler('inflation_world_bank')

In [92]:
data.loc[data['inflation_world_bank'].isna()]

Empty DataFrame
Columns: [country_name, pandem, pandem_dis, panback, country_text_id, continent, sub-region, income_group, year, rule, region_pol, region_geo, property, polyarchy, justice, education, population_density_world_bank, gdp_pc_world_bank, gini_index_world_bank, inflation_world_bank, military_world_bank, mineral_world_bank, oil_world_bank, ethnic_frac, ling_frac, relig_frac, trade_gdp_world_bank]
Index: []

[0 rows x 27 columns]

In [93]:
data.loc[data['military_world_bank'].isna()]['country_name'].to_list()

['Eritrea', 'North Korea', 'Somalia', 'Taiwan', 'Turkmenistan']

In [94]:
mean_filler('military_world_bank')

In [95]:
data.loc[data['military_world_bank'].isna()]

Empty DataFrame
Columns: [country_name, pandem, pandem_dis, panback, country_text_id, continent, sub-region, income_group, year, rule, region_pol, region_geo, property, polyarchy, justice, education, population_density_world_bank, gdp_pc_world_bank, gini_index_world_bank, inflation_world_bank, military_world_bank, mineral_world_bank, oil_world_bank, ethnic_frac, ling_frac, relig_frac, trade_gdp_world_bank]
Index: []

[0 rows x 27 columns]

In [96]:
data.loc[data['trade_gdp_world_bank'].isna()]['country_name'].to_list()

['North Korea', 'Papua New Guinea', 'Somalia', 'Syria', 'Taiwan', 'Yemen']

In [97]:
mean_filler('trade_gdp_world_bank')

In [98]:
data.loc[data['trade_gdp_world_bank'].isna()]

Empty DataFrame
Columns: [country_name, pandem, pandem_dis, panback, country_text_id, continent, sub-region, income_group, year, rule, region_pol, region_geo, property, polyarchy, justice, education, population_density_world_bank, gdp_pc_world_bank, gini_index_world_bank, inflation_world_bank, military_world_bank, mineral_world_bank, oil_world_bank, ethnic_frac, ling_frac, relig_frac, trade_gdp_world_bank]
Index: []

[0 rows x 27 columns]

In [99]:
data.isna().sum()

country_name                      0
pandem                            0
pandem_dis                        0
panback                           0
country_text_id                   0
continent                         0
sub-region                        0
income_group                      0
year                              0
rule                              0
region_pol                        0
region_geo                        0
property                          0
polyarchy                         0
justice                           0
education                        20
population_density_world_bank     0
gdp_pc_world_bank                 0
gini_index_world_bank             0
inflation_world_bank              0
military_world_bank               0
mineral_world_bank                0
oil_world_bank                    0
ethnic_frac                       0
ling_frac                         0
relig_frac                        0
trade_gdp_world_bank              0
dtype: int64

In [100]:
data.head()

country_name    pandem  pandem_dis   panback country_text_id continent  \
0  Afghanistan  0.058824        0.05  0.000000             AFG      Asia   
1      Albania  0.176471        0.15  0.173110             ALB    Europe   
2      Algeria  0.294118        0.35  0.123340             DZA    Africa   
3       Angola  0.117647        0.10  0.041671             AGO    Africa   
4    Argentina  0.058824        0.05  0.054786             ARG  Americas   

                        sub-region         income_group  year   rule  ...  \
0                    Southern Asia           Low income  2019  0.341  ...   
1                  Southern Europe  Upper middle income  2019  0.596  ...   
2                  Northern Africa  Lower middle income  2019  0.217  ...   
3               Sub-Saharan Africa  Lower middle income  2019  0.406  ...   
4  Latin America and the Caribbean  Upper middle income  2019  0.757  ...   

   gdp_pc_world_bank  gini_index_world_bank  inflation_world_bank  \
0         502.115487                   27.8              2.302373   
1        5352.857411                   33.2              1.411091   
2        3948.343279                   27.6              1.951768   
3        2973.591160                   51.3             17.145320   
4       10006.148974                   41.4             53.548304   

   military_world_bank  mineral_world_bank  oil_world_bank  ethnic_frac  \
0             0.984561            0.007421        0.001255       0.7693   
1             1.178901            0.146426        1.365752       0.2204   
2             5.271414            0.046717       12.313697       0.3394   
3             1.777138            0.000000       15.753913       0.7867   
4             0.854561            0.308310        0.735824       0.2550   

   ling_frac  relig_frac  trade_gdp_world_bank  
0     0.6141      0.2717             45.621177  
1     0.0400      0.4719             76.728300  
2     0.4427      0.0100             52.044775  
3     0.7870      0.6276             66.378012  
4     0.0620      0.2236             32.393666  

[5 rows x 27 columns]

In [101]:
data.rename(columns={'sub-region': 'sub_region'}, inplace = True)
data = data[[
    'country_name','country_text_id', # country name/id
    'continent', 'sub_region', 'region_pol', 'region_geo', 'income_group', # classification groups
    'pandem', 'pandem_dis', 'panback', # pandemic data
    'rule', 'property', 'justice', # rule of law (vdem)
    'polyarchy', 'education', # other variables (vdem)
    'gini_index_world_bank', 'population_density_world_bank', 'gdp_pc_world_bank', 'inflation_world_bank', 
    'military_world_bank', 'oil_world_bank', 'mineral_world_bank', 'trade_gdp_world_bank', # World Bank variables
    'ethnic_frac', 'ling_frac', 'relig_frac' # Fractionalization variables
 ]]

In [102]:
data.to_csv('../../data/main_research/processed_data/final_dataset.csv', index=False)
data.to_csv('../../data/robustness_check/raw_data/dataset_main_research.csv', index=False)

In [103]:
data.head()

country_name country_text_id continent                       sub_region  \
0  Afghanistan             AFG      Asia                    Southern Asia   
1      Albania             ALB    Europe                  Southern Europe   
2      Algeria             DZA    Africa                  Northern Africa   
3       Angola             AGO    Africa               Sub-Saharan Africa   
4    Argentina             ARG  Americas  Latin America and the Caribbean   

   region_pol  region_geo         income_group    pandem  pandem_dis  \
0           8          14           Low income  0.058824        0.05   
1           1           3  Upper middle income  0.176471        0.15   
2           3           5  Lower middle income  0.294118        0.35   
3           4           7  Lower middle income  0.117647        0.10   
4           2          18  Upper middle income  0.058824        0.05   

    panback  ...  population_density_world_bank  gdp_pc_world_bank  \
0  0.000000  ...                      56.937760         502.115487   
1  0.173110  ...                     104.612263        5352.857411   
2  0.123340  ...                      17.730075        3948.343279   
3  0.041671  ...                      24.713052        2973.591160   
4  0.054786  ...                      16.258510       10006.148974   

   inflation_world_bank  military_world_bank  oil_world_bank  \
0              2.302373             0.984561        0.001255   
1              1.411091             1.178901        1.365752   
2              1.951768             5.271414       12.313697   
3             17.145320             1.777138       15.753913   
4             53.548304             0.854561        0.735824   

   mineral_world_bank  trade_gdp_world_bank  ethnic_frac  ling_frac  \
0            0.007421             45.621177       0.7693     0.6141   
1            0.146426             76.728300       0.2204     0.0400   
2            0.046717             52.044775       0.3394     0.4427   
3            0.000000             66.378012       0.7867     0.7870   
4            0.308310             32.393666       0.2550     0.0620   

   relig_frac  
0      0.2717  
1      0.4719  
2      0.0100  
3      0.6276  
4      0.2236  

[5 rows x 26 columns]

In [104]:
data.shape

(146, 26)

### Variables description:
##### Country name/id and location:
* 'country_name'
* 'country_text_id'
* 'continent' - 5 categories ('Asia', 'Europe', 'Africa', 'Americas', 'Oceania')
* 'sub-region' - 15 categories (more advanced division than in the case of continents - e.g. 'Northern Europe')

##### V-dem covid data:
* 'pandem' - is the sum of restrictions (6 types of violations - with 4 levels (0 for non, 1 for minor, 2 for some, 3 for major violations) divided by all violations recorded (maximum - 17). In our case pandem ranges from 0 to 9 (maximum: 9 out of 17 violations = 0.5294).
* 'pandem_dis' - 'pandem_dis' is builded on the 'pandem' index. 'pandem_dis' variable is the sum of restrictions (6 types of violations - with 4 levels (0 for non, 1 for minor, 2 for some, 3 for major violations) + government disinformation indicator ('govmis') with 3 levels (0 for never, 1 for occasionally, 2 for often, 3 for always or almost always) divided by the total additive score (maximum - 20 (17 for pandem and 3 for govmis)). In our case 'pandem_dis' ranges from 0 to 10 (maximum: 10 out of 20 = 0.5).
* 'panback' - is builded on the 'pandem_alt' index (which is the alternative version of 'pandem' index) and 'v2x_libdem' index (Liberal democracy index) from 2019 year for each country. In our case 'panback' has 112 unique values, so we treat 'panback' as a fully continuous variable. "The Panback index captures such inverted U-shaped vulnerability to backsliding by weighing more heavily violations recorded in countries in the mid-range of V-Dem's liberal democracy index".


##### World Bank classification (income group)
* 'income_group' - according to World Bank classification, allocation countries to one of the four groups: 'Low income', 'Upper middle income', 'Lower middle income' and 'High income'.

##### V-dem dataset:
* 'rule' ('v2x_rule' in V-dem) - "To what extent are laws transparently, independently, predictably, impartially, and equally enforced, and to what extent do the actions of government officials comply with the law?". Interval, from low to high (0-1).
* 'justice' ('v2xcl_acjst' in V-dem) - "Do citizens enjoy secure and effective access to justice?". Interval, from low to high (0-1).
* 'property' ('v2xcl_prpty' in V-dem) - "Do citizens enjoy the right to private property?". Interval, from low to high (0-1).
* 'polyarchy' ('v2x_polyarchy' in V-dem) - Electoral democracy index. "To what extent is the ideal of electoral democracy in its fullest sense achieved?". Interval, from low to high (0-1).
* 'education' ('e_peaveduc' in V-dem) - "What is the average years of education among citizens older than 15?".
* 'region_pol' ('e_regionpol' in V-dem) - "In which politico-geographic region is this country located?". (8 levels: 1. Eastern Europe and post Soviet Union (including Central Asia, Mongolia); 2. Latin America; 3. North Africa and the Middle East; 4. Sub–Saharan Africa; 5. Western Europe and North America (including Australia, New Zealand); 6. East Asia (including Japan, excluding Mongolia); 7. South–East Asia; 8. South Asia.)
* 'region_geo' ('e_regiongeo' in V-dem) - "In which geographic region is this country located?". (19 levels).


##### World Bank database:
* 'Gini' - "Gini index measures the extent to which the distribution of income (or, in some cases, consumption expenditure) among individuals or households within an economy deviates from a perfectly equal distribution."
* 'Density' - "Population density is midyear population divided by land area in square kilometers."
* 'GDPpc' - "GDP per capita is gross domestic product divided by midyear population. Data are in current U.S. dollars."
* 'Inflation' - "Inflation as measured by the consumer price index reflects the annual percentage change in the cost to the average consumer of acquiring a basket of goods and services that may be fixed or changed at specified intervals, such as yearly. The Laspeyres formula is generally used."
* 'Military' - military expenditures data (% of GDP).
* 'Oil' - "Oil rents are the difference between the value of crude oil production at world prices and total costs of production." (% of GDP).
* 'Mineral' - "Mineral rents are the difference between the value of production for a stock of minerals at world prices and their total costs of production. Minerals included in the calculation are tin, gold, lead, zinc, iron, copper, nickel, silver, bauxite, and phosphate." (% of GDP).
* 'Trade_GDP' - sum of exports and imports of goods and services (% of GDP).


##### Fractionalization data (based on Alesina et al's analysis, 2003):
Fractionalization measures are computing the probability that two randomly drawn individuals are not from the same group :
* 'ethnic_frac' - ethnic fractionalization
* 'ling_frac' - linguistic fractionalization
* 'relig_frac' - religious fractionalization